lib/puppet/provider/user/useradd.rb

```diff
@@ -203,7 +203,10 @@ def deletecmd
     else
       cmd = [command(:delete)]
     end
-    cmd += @resource.managehome? ? ['-r'] : []
+    # Solaris `userdel -r` will fail if the homedir does not exist.
+    if @resource.managehome? and ('Solaris' != Facter.value(:operatingsystem) or Dir.exist?(Dir.home(@resource[:name])))
```

Comment:
nit, could you use boolean `&&` and `||`. Ruby's `and` and `or` can behave in surprising ways due to operator precedence. It doesn't matter in this particular case, but we try to avoid adding new code with it.

Reply:
@joshcooper -- Changed as requested; the operator precedence should be crystal clear, now.

Should I make similar changes in other files?